torch/nn/parallel/distributed.py

```diff
@@ -242,7 +274,13 @@ def _queue_reduction(self, bucket_idx):
 
         # Queue the reduction and make sure backward waits for it
         event = threading.Event()
-        self._reduction_queues[bucket_idx].put((dev_buckets, dev_events, event))
+
+        if dist._backend == "nccl":
+            # NCCL backend, all buckets will share a single reduction queue
+            self._reduction_queues[0].put((dev_buckets, dev_events, event))
```

Comment:
@csarofeen Every reduction bucket will write to this single queue when any of the bucket is full (in line 257). Among all processes on different nodes, the order for each bucket to become full is deterministic since the way we divide the gradients into bucket is the same across multiple processes. Therefore, the single queue and single reduction thread makes sure that the order in which each bucket got all_reduced is the same for each process.

Reply:
"the order for each bucket to become full is deterministic" This is not true. Even though your buckets are made up of all the same gradients, backwards calls are asynchronous which means the order in which the gradients finish are not deterministic. If gradient 3 is submitted before gradient 4 but gradient 3 takes much less time to compute, gradient 4 can finish before gradient 3.
I believe logic needs to be added to enforce the all_reduce of `bucket_idx-1` has been submitted before the all_reduce of `bucket_idx`.

Reply:
Sorry I take this comment back. You're right. Since autograd is submitting gpu work on a single stream/gpu this should be sufficient. Thank you for making the changes! I will pull these commits and test it out.

Reply:
A nicer way to implement this would be `queue_idx = 0 if _backend == "nccl" else bucket_idx` and then use this to queue

Reply:
Also, I can't see how this makes the order deterministic. Currently, we have no guarantees about the ordering of ops in backward, which means that reduction of 4th block can be queued before 3rd block on one node, while on the other 3rd block will be queued before the 4th (all are queued on this single queue).

Reply:
Since backward goes to the same stream on the GPU, why still no this guarantee?